backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/loops/ProgressionInfo.kt

```diff
@@ -0,0 +1,203 @@
+        parameterRestriction(0) { it.type.classifierOrNull in progressionElementClasses }
+        fqNameRestriction { it == untilFqName }
+    }
+
+    private val rangeToMatcher = createFunctionMatcher {
+        functionKind = IrFunctionMatcher.Kind.METHOD
+        receiverRestriction { it.type.classifierOrNull in progressionElementClasses }
+        fqNameRestriction { it.pathSegments().last() == rangeToName }
+        parametersSizeRestriction { it == 1 }
+        parameterRestriction(0) { it.type.classifierOrNull in progressionElementClasses }
+    }
+
+    private val downToMatcher = createFunctionMatcher {
+        functionKind = IrFunctionMatcher.Kind.EXTENSION
+        receiverRestriction { it.type.classifierOrNull in progressionElementClasses }
+        fqNameRestriction { it == downToFqName }
+        parametersSizeRestriction { it == 1 }
+        parameterRestriction(0) { it.type.classifierOrNull in progressionElementClasses }
+    }
+
+    private val stepMatcher = createFunctionMatcher {
+        functionKind = IrFunctionMatcher.Kind.EXTENSION
+        receiverRestriction { it.type.classifierOrNull in symbols.progressionClasses }
+        fqNameRestriction { it == stepFqName }
+        parametersSizeRestriction { it == 1 }
+        parameterRestriction(0) { it.type.isInt() || it.type.isLong() }
+    }
+
+    // TODO: Process constructors and other factory functions.
+    private val callHandlers = listOf(
+            indicesMatcher::match to ::buildIndices,
+            rangeToMatcher::match to ::buildRangeTo,
+            untilMatcher::match   to ::buildUntil,
+            downToMatcher::match  to ::buildDownTo,
+            stepMatcher::match    to ::buildStep
+    )
+
+    private fun handleCall(call: IrCall, progressionType: ProgressionType): ProgressionInfo? =
+            callHandlers.firstOrNull { (checker, _) ->
+                checker(call.symbol.owner)
+            }?.let { (_, builder) -> builder(call, progressionType) }
+
+    private fun buildIndices(expression: IrCall, progressionType: ProgressionType): ProgressionInfo? {
+        val int0 = IrConstImpl.int(expression.startOffset, expression.endOffset, context.irBuiltIns.intType, 0)
+
+        val bound = with(context.createIrBuilder(expression.symbol, expression.startOffset, expression.endOffset)) {
+            val clazz = expression.extensionReceiver!!.type.classifierOrFail
+            val symbol = symbols.arrayLastIndex[clazz] ?: return null
+            irCall(symbol).apply {
+                extensionReceiver = expression.extensionReceiver
+            }
+        }
+        return ProgressionInfo(progressionType, int0, bound)
+    }
+
+    private fun buildRangeTo(expression: IrCall, progressionType: ProgressionType) =
+            ProgressionInfo(progressionType,
+                    expression.dispatchReceiver!!,
+                    expression.getValueArgument(0)!!)
+
+    private fun buildUntil(expression: IrCall, progressionType: ProgressionType) =
```

Comment:
By the way, the implementation of this function duplicates some details of the corresponding matcher. So it makes sense to organize pattern matching in such way that matcher and handler are parts of the same indivisible entity. E.g.


```
UntilMatch : CallWithReceiverAndOneParameter {
    // some details about restrictions
    // ...

    override fun handle(receiver: IrExpression, argument: IrFunction)  = ProgressionInfo(...)
}
```

or even
```
UntilMatch : CallWithReceiverAndOneParameter {
    override fun handle(receiver: IrExpression, argument: IrFunction) = if (check receiver and check argument) {
        Matched(ProgressionInfo(...))
    } else {
        NotMatched
    }
}
```